[EMCAL-507] Adding basic EMCAL reconstruction workflow

mfasDa opened the pull request:
Basic EMCAL reconstruction workflow sending digits
obtained from a reader to a digits printer and
a digits writer. More components will be added
step-by-step.

The workflow is consisting of
- Publisher Spec
- Digits printer
- Writer spec
- Reconstruction workflow steering

mfasDa commented:
Basic EMCAL reco workflow consisting of a digits publisher, a tree writer and a digits printer. Adapted from the TPC reconstruction workflow. Will be extended with clusterizer / cluster writer / cell writer in the future.

Note: Some of the components (Publisher) are needed for the QC development, therefore we need this basic workflow to be merged.

matthiasrichter commented on Detectors/EMCAL/workflow/src/PublisherSpec.cxx:
```diff
@@ -0,0 +1,149 @@
+      auto publish = [&processAttributes, &pc, propagateMC]() {
+        o2::emcal::EMCALBlockHeader::PayloadType_t payloadType(o2::emcal::EMCALBlockHeader::PayloadType_t::kNoPayload);
+        LOG(DEBUG) << "[publish] Data type:" << processAttributes->datatype;
+        if (processAttributes->datatype.find("EMCALDigit") != std::string::npos) {
```
I would consider this if statement a pure debug feature which should be removed. Also, there is no need to propagate the payload type in a custom header because the data header is intended for this information. What you have to make sure in the invocation of the `getPublisherConf` method is to use the correct DataDescription.

Edit: I have now realized that you are using the payload type of the EMCALBlockHeader to propagate the end-of-data condition. I suggest to rename it, and we will find a generic solution in DPL.

mfasDa replied:
Debug printf will be removed before merging. For the field in the header I move it to simple bool.  For the moment we need the header, otherwise components in particular in the QC will break.

matthiasrichter commented on Detectors/EMCAL/workflow/src/RecoWorkflow.cxx:
```diff
@@ -0,0 +1,155 @@
+                                   "o2sim",
+                                   BranchDefinition<DigitOutputType>{ o2::framework::InputSpec{ "data", "EMC", "DIGITS", 0 },
+                                                                      "EMCDigit",
+                                                                      "digitmc-branch-name" },
```
```suggestion
                                                                      "digit-branch-name" },
```
suppose that's a typo

mfasDa replied:
Correct - changed! Thanks for spotting!

matthiasrichter commented on Detectors/EMCAL/workflow/src/RecoWorkflow.cxx:
```diff
@@ -0,0 +1,155 @@
+    // or one branch definition
+    if (propagateMC) {
+      return std::move(o2::framework::MakeRootTreeWriterSpec(processName, defaultFileName, defaultTreeName,
+                                                             o2::framework::MakeRootTreeWriterSpec::TerminationCondition{ checkReady },
```
```suggestion
                                                             1, // default number of events
```

mfasDa replied:
Shouldn't the publisher be able to publish more than just 1 event? I would prefer being able to send a variable number of events, ideally as much as in the tree.

matthiasrichter replied:
you always have the possibility to set the number of events with the `nevents` command line option

matthiasrichter commented on Detectors/EMCAL/workflow/src/RecoWorkflow.cxx:
```diff
@@ -0,0 +1,155 @@
+  // in this workflow, the EOD is sent after the last real data, and all inputs will receive EOD,
+  // so it is enough to check on the first occurence
+  // FIXME: this will be changed once DPL can propagate control events like EOD
+  auto checkReady = [](o2::framework::DataRef const& ref, bool& isReady) {
```
I suggest to remove this completely because this particular case should be solved generically in the framework.

mfasDa replied:
For the moment it is needed as other components in particular in the QC need to check this.

matthiasrichter replied:
ok, fair enough, so you want to merge this workflow with another one, but in this case you don't need the writer at all. However, I'm fine with that approach.

There will be a conflict with #2124 because I modified the return value of the evaluator functor.

mfasDa replied:
I removed the isReady argument. As local variable it is not needed.

matthiasrichter commented on Detectors/EMCAL/workflow/src/RecoWorkflow.cxx:
```diff
@@ -0,0 +1,155 @@
+                                                             std::move(mcbranch)));
+    }
+    return std::move(o2::framework::MakeRootTreeWriterSpec(processName, defaultFileName, defaultTreeName,
+                                                           o2::framework::MakeRootTreeWriterSpec::TerminationCondition{ checkReady },
```
same here

mfasDa commented:
Hello @matthiasrichter ,

Thanks a lot for the comment! For what concerns the removal of the block header for the terminate condition: I use the publisher also in a dedicated test workflow for EMCAL for the development of the QC components. The QC task is crashing as well with the  TMessage deserialization exception if I don't handle the terminate condition. I am not sure how the termination is handled there (adding @Barthelemy). Before removing the termination condition I would like to make sure I am not breaking anything in the our QC test workflow.

Cheers

Markus

mfasDa commented:
Hello @matthiasrichter ,

We might have a need for the header different than the terminate condition, which is propagating the trigger type. EMCAL will still be triggered in run3, and providing access to the trigger type will be particularly of relevance for the QC as distributions show differences for different triggers. Are basic event information (i.e. timestamp, timeframe, or trigger as in case of EMCAL) made available to the processors by the DPL i.e. in a data header or is it our responsibility to take care of it?

Cheers

Markus

mfasDa commented on DataFormats/Detectors/EMCAL/src/EMCALBlockHeader.cxx:
```diff
@@ -0,0 +1,13 @@
+
+#include "DataFormatsEMCAL/EMCALBlockHeader.h"
+
+constexpr o2::header::HeaderType o2::emcal::EMCALBlockHeader::sHeaderType;
```
At the moment the Block Header is the first class in DataFormatsEMCAL, so in order to make the library compiling I need a cxx file. But there are other classes to come (i.e. clusters, cells), so once they are merged I can remove the cxx file.

For the second are you talking about this?
```
constexpr static const o2::header::HeaderType sHeaderType = "EMCBlkH";
```

matthiasrichter replied:
ok, and yes, the header type member can be set like this

mfasDa replied:
Fixed

mfasDa commented on DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/EMCALBlockHeader.h:
```diff
@@ -0,0 +1,43 @@
+    kDigits,          ///< Digits payload
+    kClusters,        ///< Clusters payload
+    kRaw              ///< Raw data payload
+  };
```
Will be replaced by simple bool.

mfasDa commented:
Before merging I would like to squash the commits first.

matthiasrichter commented on DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/EMCALBlockHeader.h:
```diff
@@ -0,0 +1,36 @@
+/// @brief Header for EMCAL flagging the following EMCAL payload
+struct EMCALBlockHeader : public o2::header::BaseHeader {
+  // Required to do the lookup
+  constexpr static const o2::header::HeaderType sHeaderType = "EMCBlkH";
```
since there is the cxx file with the static variable you should do the assignment there

mfasDa replied:
Done

mfasDa replied:
Leading to a compiler error under gcc 7.3 (not under clang on mac):

mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/EMCALBlockHeader.h:25:49: error: constexpr static data member 'sHeaderType' must have an initializer
   constexpr static const o2::header::HeaderType sHeaderType;
                                                 ^~~~~~~~~~~

matthiasrichter commented on DataFormats/Detectors/EMCAL/src/EMCALBlockHeader.cxx:
```diff
@@ -0,0 +1,13 @@
+
+#include "DataFormatsEMCAL/EMCALBlockHeader.h"
+
+constexpr o2::header::HeaderType o2::emcal::EMCALBlockHeader::sHeaderType;
```
newline missing

matthiasrichter commented on Detectors/EMCAL/workflow/include/EMCALWorkflow/DigitsPrinterSpec.h:
```diff
@@ -0,0 +1,26 @@
+}
+} // namespace emcal
+
+} // namespace o2
```
newline missing

mfasDa replied:
Done

matthiasrichter commented on Detectors/EMCAL/workflow/include/EMCALWorkflow/PublisherSpec.h:
```diff
@@ -0,0 +1,42 @@
+framework::DataProcessorSpec getPublisherSpec(PublisherConf const& config, bool propagateMC = true);
+
+} // namespace emcal
+} // end namespace o2
```
newline missing

mfasDa replied:
Done

matthiasrichter commented on Detectors/EMCAL/workflow/src/DigitsPrinterSpec.cxx:
```diff
@@ -0,0 +1,53 @@
+                                           { { "digits", o2::header::gDataOriginEMC, "DIGITS", 0, o2::framework::Lifetime::Timeframe } },
+                                           {},
+                                           o2::framework::AlgorithmSpec(initFunction) };
+}
```
newline missing

mfasDa replied:
Done

matthiasrichter commented:
You can either squash yourself using interactive rebase and than force push the branch, or we do it when merging the PR. If you are done with the changes you can remove the WIP tag to signal that it's ready for final review.

matthiasrichter commented:
> Hello @matthiasrichter ,
> 
> Thanks a lot for the comment! For what concerns the removal of the block header for the terminate condition: I use the publisher also in a dedicated test workflow for EMCAL for the development of the QC components. The QC task is crashing as well with the TMessage deserialization exception if I don't handle the terminate condition. I am not sure how the termination is handled there (adding @Barthelemy). Before removing the termination condition I would like to make sure I am not breaking anything in the our QC test workflow.
> 
> Cheers
> 
> Markus

As far as concerning the exception your are mentioning, this you can simply avoid by not sending a non-Root-serialized message in the publisher. As all these workflows are simple tests for the moment, I would keep them simple and skip the workarounds for implementing the automatic termination. I had a chat with Giulio how to handle and propagate such control information like an "end-of-data" condition.

matthiasrichter commented on Detectors/EMCAL/workflow/src/PublisherSpec.cxx:
```diff
@@ -0,0 +1,134 @@
+        if (propagateMC) {
+          pc.outputs().snapshot(o2::framework::OutputRef{ "outputMC", 0, { dummyheader } }, 0);
+        }
+      }
```
If you drop the handling of the no-more-data case, you avoid the exception in the receiving process.

matthiasrichter commented on Detectors/EMCAL/workflow/src/RecoWorkflow.cxx:
```diff
@@ -0,0 +1,145 @@
+  // in this workflow, the EOD is sent after the last real data, and all inputs will receive EOD,
+  // so it is enough to check on the first occurence
+  // FIXME: this will be changed once DPL can propagate control events like EOD
+  auto checkReady = [](o2::framework::DataRef const& ref, bool& isReady) {
```
```suggestion
  auto checkReady = [](o2::framework::DataRef const& ref) {
  bool isReady = false;
```

mfasDa replied:
FIxed

matthiasrichter commented on Detectors/EMCAL/workflow/src/RecoWorkflow.cxx:
```diff
@@ -0,0 +1,145 @@
+        return o2::framework::MakeRootTreeWriterSpec::TerminationCondition::Action::SkipProcessing;
+      }
+    }
+    return o2::framework::MakeRootTreeWriterSpec::TerminationCondition::Action::DoProcessing;
```
```suggestion
    return std::make_tuple(MakeRootTreeWriterSpec::TerminationCondition::Action::DoProcessing, isReady);
```

mfasDa replied:
Fixed

matthiasrichter commented on Detectors/EMCAL/workflow/src/RecoWorkflow.cxx:
```diff
@@ -0,0 +1,145 @@
+      isReady = !emcalheader->mHasPayload;
+      // indicate normal processing if not ready and skip if ready
+      if (isReady) {
+        return o2::framework::MakeRootTreeWriterSpec::TerminationCondition::Action::SkipProcessing;
```
```suggestion
        return std::make_tuple(o2::framework::MakeRootTreeWriterSpec::TerminationCondition::Action::SkipProcessing, isReady);
```

mfasDa replied:
Fixed

Barthelemy commented:
Hello, 
Sorry, I completely overlooked your call. Is my help still needed here ? 
Cheers,
Barth

mfasDa commented:
Hi @Barthelemy ,

Yes, this is related to https://github.com/AliceO2Group/QualityControl/pull/198: What is the terminate condition for components in the QC workflows? Because I am currently using the publisher from this workflow in order to supply the QC workflow with data, and I get a crash when not checking for the header in this workflow, and just want to make sure with sending a fixed amount of events the workflow doesn't crash. Second item here: EMCAL is triggered and we need to fill several histograms in the QC as function of the trigger class. Is the trigger class somewhere available in framework headers accessible by the QC? If not it could be propagated via the EMCAL header.

Cheers

Markus

mfasDa commented:
Hi @matthiasrichter ,

Thanks a lot for your comments! I have pushed a fix addressing the issues you pointed out. Is the code now ready to merge?

Cheers

Markus

mfasDa commented:
@matthiasrichter I have adapted the PublisherSpec to the latest changes in the RootTreeReader. Anything else that needs to be done or is the PR ready to be merged?
